aigerbv/common.py

```diff
@@ -81,6 +80,22 @@ def bitwise_negate(wordlen, input='x', output='not x'):
     )
 
 
+def reduce_binop(wordlen, inputs, output, op):
+    def join(left, right):
+        (o1, *_), (o2, *_) = left.outputs, right.outputs # noqa
+        return (left | right) >> op(wordlen, o1, o2, _fresh())
+
+    inputs = list(inputs)
+    queue = [identity_gate(wordlen, i) for i in inputs]
+    while len(queue) > 1:
+        queue = list(starmap(join, zip(queue[::2], queue[1::2])))
+
+    circ = queue[0]
+    if len(inputs) & 1:
+        circ = join(circ, identity_gate(wordlen, inputs[-1]))
+    return circ
+
+
```

Comment:
Nice, but looks a little complicated. Took me a while to understand line 91. Maybe decompress?

Can line 93 to 95 be merged into the first loop? Not sure what's going on.

Reply:
The last part is handling even vs odd case. I'm building the circuit as a balanced tree, by pairing every element. That's what line 91 does in a bit a convoluted way. What I will do is make a helper function called `_apply_pairwise` that will hopefully make it more readable.